python/tvm/topi/x86/dense_alter_op.py

```diff
@@ -39,6 +40,17 @@ def _alter_dense_layout(attrs, inputs, tinfos, out_type):
         relay.op.get("nn.dense"), attrs, tinfos, out_type, target
     )
     workload = autotvm.task.get_workload(outs)
+
+    data_type = relay.transform.InferType()(IRModule.from_expr(inputs[0]))["main"].body.checked_type
+    assert len(data_type.shape) in [2, 3], "Input data must be a 2D or 3D tensor."
+
+    if len(data_type.shape) == 3:
+        # Only supports 2D input
+        in_layout = "NC%dc" % data_type.shape[2]
+        data = relay.layout_transform(inputs[0], in_layout, "NC")
```

Comment:
I took some time to think about the overhead of implicitly inserted layout_transform when input data is NC[x]c, but I didn't have a better solution, so I guess this is the right one we're looking for.